OpenRA.Mods.RA/Widgets/ProductionPaletteWidget.cs

```diff
@@ -87,6 +90,35 @@ public ProductionPaletteWidget(OrderManager orderManager, World world, WorldRend
 			clock = new Animation(world, "clock");
 		}
 
+		public void ScrollDown()
+		{
+			var totalRows = AllBuildables.Count() / Columns;
+
+			// Check is there is an incomplete row
+			if (AllBuildables.Count() % Columns != 0)
+				totalRows++;
+
+			if (IconRowOffset < totalRows - MaxRows)
+				IconRowOffset++;
+		}
+
+		public void ScrollUp()
+		{
+			if (IconRowOffset > 0)
+				IconRowOffset--;
+		}
+
+		public IEnumerable<ActorInfo> AllBuildables
```

Comment:
We don't really want to evaluate `AllBuildables.Count()` everywhere (its not a good idea performance-wise), so you should cache both `IconCount` and `DisplayedIconCount` in the `Tick` method.